// @8base/generators/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  generateCreateForm: true,
  generateEditForm: true,
  generateDeleteForm: true,
  generateTable: true,
  generateIndex: true,
  generateRoot: true,
  generateAllScreens: true,
  generateScreen: true,
  replaceInitialApp: true
};
Object.defineProperty(exports, "generateCreateForm", {
  enumerable: true,
  get: function get() {
    return _createForm.generateCreateForm;
  }
});
Object.defineProperty(exports, "generateEditForm", {
  enumerable: true,
  get: function get() {
    return _editForm.generateEditForm;
  }
});
Object.defineProperty(exports, "generateDeleteForm", {
  enumerable: true,
  get: function get() {
    return _deleteForm.generateDeleteForm;
  }
});
Object.defineProperty(exports, "generateTable", {
  enumerable: true,
  get: function get() {
    return _table.generateTable;
  }
});
Object.defineProperty(exports, "generateIndex", {
  enumerable: true,
  get: function get() {
    return _generateIndex.generateIndex;
  }
});
Object.defineProperty(exports, "generateRoot", {
  enumerable: true,
  get: function get() {
    return _generateRoot.generateRoot;
  }
});
Object.defineProperty(exports, "generateAllScreens", {
  enumerable: true,
  get: function get() {
    return _generateAllScreens.generateAllScreens;
  }
});
Object.defineProperty(exports, "generateScreen", {
  enumerable: true,
  get: function get() {
    return _generateScreen.generateScreen;
  }
});
Object.defineProperty(exports, "replaceInitialApp", {
  enumerable: true,
  get: function get() {
    return _replaceInitialApp.replaceInitialApp;
  }
});

var _createForm = require("./generateFiles/generateCreateForm/createForm");

var _editForm = require("./generateFiles/generateEditForm/editForm");

var _deleteForm = require("./generateFiles/generateDeleteForm/deleteForm");

var _table = require("./generateFiles/generateTable/table");

var _generateIndex = require("./generateFiles/generateIndex");

var _generateRoot = require("./generateFiles/generateRoot/generateRoot");

var _generateAllScreens = require("./generateFs/generateAllScreens");

var _generateScreen = require("./generateFs/generateScreen");

var _replaceInitialApp = require("./generateFs/replaceInitialApp");

var _constants = require("./constants");

Object.keys(_constants).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _constants[key];
    }
  });
});
