// 8base-cli/dist/engine/commands/backup/commands/restore.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const translations_1 = require("../../../../common/translations");
const configuraion_1 = require("../../../../common/configuraion");
const execute_1 = require("../../../../common/execute");
const GraphqlActions_1 = require("../../../../consts/GraphqlActions");
exports.default = {
    command: 'restore',
    handler: (params, context) => __awaiter(this, void 0, void 0, function* () {
        let { environment, backup } = params;
        configuraion_1.ProjectConfigurationState.expectConfigured(context);
        context.spinner.start(context.i18n.t('backup_restore_in_progress'));
        yield execute_1.executeAsync(context, GraphqlActions_1.GraphqlAsyncActions.backupRestore, { name: environment, backup });
        context.spinner.stop();
    }),
    describe: translations_1.translations.i18n.t('backup_restore_describe'),
    builder: (args) => args
        .usage(translations_1.translations.i18n.t('backup_restore_usage'))
        .option('backup', {
        alias: 'b',
        describe: translations_1.translations.i18n.t('backup_restore_set_backup_name_describe'),
        type: 'string',
        demandOption: true,
    })
        .option('environment', {
        alias: 'e',
        describe: translations_1.translations.i18n.t('backup_restore_set_environment_describe'),
        type: 'string',
        demandOption: true,
    }),
};
