// 8base-cli/dist/engine/commands/generate/commands/mock.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const translations_1 = require("../../../../common/translations");
const projectController_1 = require("../../../controllers/projectController");
exports.default = {
    command: 'mock <name>',
    handler: (params, context) => __awaiter(this, void 0, void 0, function* () {
        let { name, mockName, silent } = params;
        projectController_1.ProjectController.generateMock(context, {
            name: mockName,
            silent,
            functionName: name,
        });
    }),
    describe: translations_1.translations.i18n.t('generate_mock_describe'),
    builder: (args) => args
        .usage(translations_1.translations.i18n.t('generate_mock_usage'))
        .option('silent', {
        describe: translations_1.translations.i18n.t('silent_describe'),
        default: false,
        type: 'boolean',
    })
        .option('mockName', {
        alias: 'm',
        describe: translations_1.translations.i18n.t('generate_mock_name_describe'),
        default: 'request',
        type: 'string',
    }),
};
