// 8base-cli/dist/engine/commands/generate/commands/trigger.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const translations_1 = require("../../../../common/translations");
const Extensions_1 = require("../../../../interfaces/Extensions");
const projectController_1 = require("../../../controllers/projectController");
exports.default = {
    command: 'trigger <name>',
    handler: (params, context) => __awaiter(this, void 0, void 0, function* () {
        let { name, type, operation, mocks, syntax, silent } = params;
        if (operation && !/[\w\d]+\.(create|update|delete)/.test(operation)) {
            throw new Error(translations_1.translations.i18n.t('generate_trigger_invalid_operation'));
        }
        projectController_1.ProjectController.generateFunction(context, {
            type: Extensions_1.ExtensionType.trigger,
            name,
            mocks,
            syntax,
            silent,
        }, {
            type,
            operation,
        });
    }),
    describe: translations_1.translations.i18n.t('generate_trigger_describe'),
    builder: (args) => args
        .usage(translations_1.translations.i18n.t('generate_trigger_usage'))
        .option('type', {
        alias: 't',
        describe: translations_1.translations.i18n.t('generate_trigger_type_describe'),
        type: 'string',
        choices: Object.values(Extensions_1.TriggerType),
    })
        .option('operation', {
        alias: 'o',
        describe: translations_1.translations.i18n.t('generate_trigger_operation_describe'),
        type: 'string',
    })
        .option('mocks', {
        alias: 'x',
        describe: translations_1.translations.i18n.t('generate_mocks_describe'),
        default: true,
        type: 'boolean',
    })
        .option('syntax', {
        alias: 's',
        describe: translations_1.translations.i18n.t('generate_syntax_describe'),
        default: 'ts',
        type: 'string',
        choices: Object.values(Extensions_1.SyntaxType),
    })
        .option('silent', {
        describe: translations_1.translations.i18n.t('silent_describe'),
        default: false,
        type: 'boolean',
    }),
};
