// @8base/generators/dist/generateFiles/generateIndex/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateIndex = void 0;

var changeCase = _interopRequireWildcard(require("change-case"));

var ejs = _interopRequireWildcard(require("ejs"));

var pluralize = _interopRequireWildcard(require("pluralize"));

var _formatCode = require("../../formatCode");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

/* babel-plugin-inline-import './index.js.ejs' */
// @ts-ignore
var index = "import React from 'react';\nimport { Card, Heading } from '@8base/boost';\n\nimport { <%= changeCase.pascal(entityName)%>CreateDialog } from './<%= changeCase.pascal(entityName)%>CreateDialog';\nimport { <%= changeCase.pascal(entityName)%>EditDialog } from './<%= changeCase.pascal(entityName)%>EditDialog';\nimport { <%= changeCase.pascal(entityName)%>DeleteDialog } from './<%= changeCase.pascal(entityName)%>DeleteDialog';\nimport { <%= changeCase.pascal(screenName)%>Table } from './<%= changeCase.pascal(screenName)%>Table';\n\nconst <%= changeCase.pascal(screenName)%> = () => (\n  <Card padding=\"md\" stretch>\n    <Card.Header>\n      <Heading type=\"h4\" text=\" <%= changeCase.pascal(screenName)%>\" />\n    </Card.Header>\n\n    <<%= changeCase.pascal(entityName)%>CreateDialog />\n    <<%= changeCase.pascal(entityName)%>EditDialog />\n    <<%= changeCase.pascal(entityName)%>DeleteDialog />\n    <Card.Body padding=\"none\" stretch scrollable>\n      <<%= changeCase.pascal(screenName)%>Table />\n    </Card.Body>\n  </Card>\n);\n\nexport { <%= changeCase.pascal(screenName)%> };\n";

var generateIndex = function generateIndex(_ref) {
  var tablesList = _ref.tablesList,
      tableId = _ref.tableId,
      screenName = _ref.screenName;
  var table = tablesList.find(function (_ref2) {
    var id = _ref2.id;
    return tableId === id;
  });

  if (!table) {
    throw new Error("Can't find a table with ".concat(tableId, " id"));
  }

  var tableName = table.displayName || table.name;
  var entityName = pluralize.singular(screenName || tableName);
  var tableGenerated = ejs.render(index, {
    changeCase: changeCase,
    entityName: entityName,
    pluralize: pluralize,
    screenName: screenName || tableName,
    tableName: tableName
  });
  return (0, _formatCode.formatCode)(tableGenerated);
};

exports.generateIndex = generateIndex;
