// 8base-cli/dist/engine/commands/project/commands/info.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
const translations_1 = require("../../../../common/translations");
const configuraion_1 = require("../../../../common/configuraion");
const Environment_1 = require("../../../../consts/Environment");
exports.default = {
    command: 'info',
    handler: (params, context) => __awaiter(this, void 0, void 0, function* () {
        configuraion_1.ProjectConfigurationState.expectConfigured(context);
        const { workspaceId, environmentName } = context.workspaceConfig;
        const workspace = (yield resolveWorkspace(context, workspaceId)) || {
            apiHost: Environment_1.DEFAULT_REMOTE_ADDRESS,
            id: workspaceId,
            name: '',
        };
        context.logger.info(translations_1.translations.i18n.t('project_info_text', {
            workspaceId: chalk_1.default.green(workspaceId),
            workspaceName: chalk_1.default.green(workspace.name),
            environment: chalk_1.default.green(environmentName),
            endpoint: chalk_1.default.green(`${workspace.apiHost}/${workspaceId}`),
        }));
    }),
    describe: translations_1.translations.i18n.t('project_info_describe'),
    builder: (args) => args.usage(translations_1.translations.i18n.t('project_info_usage')),
};
const resolveWorkspace = (context, workspaceId) => __awaiter(this, void 0, void 0, function* () {
    try {
        return (yield context.getWorkspaces()).find(({ id }) => id === workspaceId);
    }
    catch (e) {
        // in case of api token user could not resolve workspace list.
    }
    return null;
});
