// 8base-cli/dist/config/static.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const predefineData_1 = require("./predefineData");
class StaticConfig {
    static get projectTemplatePath() {
        return this.staticData.projectTemplatePath;
    }
    static get functionTemplatesPath() {
        return this.staticData.functionTemplatesPath;
    }
    static get pluginTemplatePath() {
        return this.staticData.pluginTemplatePath;
    }
    static get mockTemplatePath() {
        return this.staticData.mockTemplatePath;
    }
    static get rootProjectDir() {
        return this.staticData.projectDir;
    }
    static get rootExecutionDir() {
        return this.staticData.executionDir;
    }
    static get commandsDir() {
        return this.staticData.commandsPath;
    }
    static get homePath() {
        return process.env.USERPROFILE || process.env.HOME || process.env.HOMEPATH;
    }
    static get authDomain() {
        return this.staticData.authDomain;
    }
    static get authClientId() {
        return this.staticData.authClientId;
    }
    static get webClientAddress() {
        return this.staticData.webClientAddress;
    }
    static get serviceConfigFileName() {
        return path.join(this.staticData.executionDir, '8base.yml');
    }
    static get packageFileName() {
        return 'package.json';
    }
}
StaticConfig.staticData = new predefineData_1.PredefineData();
/**
 * Compiler paths
 */
StaticConfig.buildRootFolder = '.build';
StaticConfig.buildDistFolder = 'dist';
StaticConfig.modulesFolder = 'node_modules';
StaticConfig.metaFolder = 'meta';
StaticConfig.packageFolder = 'package';
StaticConfig.buildRootDirPath = path.join(StaticConfig.rootExecutionDir, StaticConfig.buildRootFolder);
StaticConfig.buildDistPath = path.join(StaticConfig.buildRootDirPath, StaticConfig.buildDistFolder);
StaticConfig.metaDir = path.join(StaticConfig.buildRootDirPath, StaticConfig.metaFolder);
StaticConfig.packageDir = path.join(StaticConfig.buildRootDirPath, StaticConfig.packageFolder);
StaticConfig.FunctionHandlerExt = '.js';
StaticConfig.supportedCompileExtension = new Set(['.ts', '.js']);
exports.StaticConfig = StaticConfig;
