// 8base-cli/dist/interfaces/Extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ExtensionType;
(function (ExtensionType) {
    ExtensionType["resolver"] = "resolver";
    ExtensionType["trigger"] = "trigger";
    ExtensionType["webhook"] = "webhook";
    ExtensionType["task"] = "task";
})(ExtensionType = exports.ExtensionType || (exports.ExtensionType = {}));
var SyntaxType;
(function (SyntaxType) {
    SyntaxType["js"] = "js";
    SyntaxType["ts"] = "ts";
})(SyntaxType = exports.SyntaxType || (exports.SyntaxType = {}));
var DeployModeType;
(function (DeployModeType) {
    DeployModeType["full"] = "FULL";
    DeployModeType["plugins"] = "ONLY_PLUGINS";
    DeployModeType["project"] = "ONLY_PROJECT";
    DeployModeType["functions"] = "FUNCTIONS";
    DeployModeType["migrations"] = "MIGRATIONS";
})(DeployModeType = exports.DeployModeType || (exports.DeployModeType = {}));
var GraphQLFunctionType;
(function (GraphQLFunctionType) {
    GraphQLFunctionType["Mutation"] = "Mutation";
    GraphQLFunctionType["Query"] = "Query";
})(GraphQLFunctionType = exports.GraphQLFunctionType || (exports.GraphQLFunctionType = {}));
var TriggerOperation;
(function (TriggerOperation) {
    TriggerOperation["create"] = "create";
    TriggerOperation["update"] = "update";
    TriggerOperation["delete"] = "delete";
})(TriggerOperation = exports.TriggerOperation || (exports.TriggerOperation = {}));
var TriggerType;
(function (TriggerType) {
    TriggerType["before"] = "before";
    TriggerType["after"] = "after";
})(TriggerType = exports.TriggerType || (exports.TriggerType = {}));
