// @8base/generators/dist/generateFiles/generateDeleteForm/deleteForm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateDeleteForm = void 0;

var _schemaNameGenerator = require("@8base/schema-name-generator");

var changeCase = _interopRequireWildcard(require("change-case"));

var ejs = _interopRequireWildcard(require("ejs"));

var pluralize = _interopRequireWildcard(require("pluralize"));

var _formatCode = require("../../formatCode");

var _chunks = require("../chunks");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

/* babel-plugin-inline-import './deleteForm.js.ejs' */
// @ts-ignore
var deleteForm = "import React from 'react';\nimport gql from 'graphql-tag';\nimport { Form as FormLogic } from '@8base/forms';\nimport { Dialog, Button, ModalContext } from '@8base/boost';\nimport { graphql } from 'react-apollo';\n\nconst <%= changeCase.upper(entityName)%>_DELETE_MUTATION = gql`\n  mutation <%= changeCase.pascal(entityName)%>Delete($id: ID!) {\n    <%= SchemaNameGenerator.getDeleteItemFieldName(tableName)%>(data: { id: $id }) {\n      success\n    }\n  }\n`;\n\nconst enhancer = graphql(<%= changeCase.upper(entityName)%>_DELETE_MUTATION, {\n  name: '<%= SchemaNameGenerator.getDeleteItemFieldName(tableName)%>',\n  options: {\n    refetchQueries: ['<%=changeCase.pascal(tableName)%>TableContent', '<%= changeCase.pascal(SchemaNameGenerator.getTableListFieldName(tableName))%>'],\n    context: {\n      TOAST_SUCCESS_MESSAGE: '<%= changeCase.pascal(entityName)%> successfully deleted',\n    },\n  },\n});\n\nconst <%= changeCase.pascal(entityName)%>DeleteDialog = enhancer(\n  class <%= changeCase.pascal(entityName)%>DeleteDialog extends React.Component {\n    static contextType = ModalContext;\n\n    createOnSubmit = id => async () => {\n      await this.props.<%= SchemaNameGenerator.getDeleteItemFieldName(tableName)%>({ variables: { id } });\n\n      this.context.closeModal('<%= changeCase.upper(entityName)%>_DELETE_DIALOG_ID');\n    };\n\n    onClose = () => {\n      this.context.closeModal('<%= changeCase.upper(entityName)%>_DELETE_DIALOG_ID');\n    };\n\n    renderFormContent = ({ handleSubmit, invalid, submitting }) => (\n      <form onSubmit={handleSubmit}>\n        <Dialog.Header title=\"Delete <%= changeCase.pascal(entityName)%>\" onClose={this.onClose} />\n        <Dialog.Body scrollable>Do you really want to delete <%= changeCase.lower(entityName)%>?</Dialog.Body>\n        <Dialog.Footer>\n          <Button color=\"neutral\" variant=\"outlined\" disabled={submitting} onClick={this.onClose}>\n            Cancel\n          </Button>\n          <Button color=\"danger\" type=\"submit\" disabled={invalid} loading={submitting}>\n            Delete <%= changeCase.pascal(entityName)%>\n          </Button>\n        </Dialog.Footer>\n      </form>\n    );\n\n    renderContent = ({ args }) => {\n      return <FormLogic onSubmit={this.createOnSubmit(args.id)} formatRelationToIds>{this.renderFormContent}</FormLogic>;\n    };\n\n    render() {\n      return (\n        <Dialog id={'<%= changeCase.upper(entityName)%>_DELETE_DIALOG_ID'} size=\"sm\">\n          {this.renderContent}\n        </Dialog>\n      );\n    }\n  }\n);\n\nexport { <%= changeCase.pascal(entityName)%>DeleteDialog };\n";

var generateDeleteForm = function generateDeleteForm(_ref) {
  var tablesList = _ref.tablesList,
      tableId = _ref.tableId,
      screenName = _ref.screenName;
  var table = tablesList.find(function (_ref2) {
    var id = _ref2.id;
    return tableId === id;
  });

  if (!table) {
    throw new Error("Can't find a table with ".concat(tableId, " id"));
  }

  var tableName = table.displayName || table.name;
  var fields = table.fields.filter(function (_ref3) {
    var isMeta = _ref3.isMeta;
    return !isMeta;
  });
  var entityName = pluralize.singular(screenName || tableName);
  var tableGenerated = ejs.render(deleteForm, {
    SchemaNameGenerator: _schemaNameGenerator.SchemaNameGenerator,
    changeCase: changeCase,
    chunks: _chunks.chunks,
    entityName: entityName,
    fields: fields,
    pluralize: pluralize,
    screenName: screenName || tableName,
    tableName: tableName
  });
  return (0, _formatCode.formatCode)(tableGenerated);
};

exports.generateDeleteForm = generateDeleteForm;
