// @8base/generators/dist/generateFs/generateScreen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateScreen = void 0;

var changeCase = _interopRequireWildcard(require("change-case"));

var os = _interopRequireWildcard(require("os"));

var pluralize = _interopRequireWildcard(require("pluralize"));

var _formatCode = require("../formatCode");

var _chunks = require("../generateFiles/chunks");

var _createForm = require("../generateFiles/generateCreateForm/createForm");

var _deleteForm = require("../generateFiles/generateDeleteForm/deleteForm");

var _editForm = require("../generateFiles/generateEditForm/editForm");

var _generateIndex = require("../generateFiles/generateIndex");

var _table = require("../generateFiles/generateTable/table");

var _generateFileNames = require("./generateFileNames");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var PAGE_CONSTANTS = {
  APP_PAGES_IMPORTS: '/** __APP_PAGES_IMPORTS__ */',
  APP_ROUTES: '{/** __APP_ROUTES__ */}',
  APP_ROUTE_LINKS: '{/** __APP_ROUTE_LINKS__ */}'
};

var generateScreen = function generateScreen(_ref, config) {
  var tablesList = _ref.tablesList,
      screen = _ref.screen,
      rootFile = _ref.rootFile;
  var fs = {};
  var screenName = screen.screenName,
      tableId = screen.tableId,
      tableFields = screen.tableFields,
      formFields = screen.formFields;
  var generatorData = {
    tablesList: tablesList,
    tableId: tableId,
    screenName: screenName
  };
  var routeUrl = "/".concat(changeCase.camel(screenName));
  var entityName = pluralize.singular(screenName);
  fs["src/routes/".concat((0, _generateFileNames.getScreenFolderName)(screenName), "/").concat((0, _generateFileNames.getCreateFormFileName)(entityName))] = (0, _createForm.generateCreateForm)(generatorData, _objectSpread({}, config, {
    includeColumns: formFields
  }));
  fs["src/routes/".concat((0, _generateFileNames.getScreenFolderName)(screenName), "/").concat((0, _generateFileNames.getEditFormFileName)(entityName))] = (0, _editForm.generateEditForm)(generatorData, _objectSpread({}, config, {
    includeColumns: formFields
  }));
  fs["src/routes/".concat((0, _generateFileNames.getScreenFolderName)(screenName), "/").concat((0, _generateFileNames.getDeleteFormFileName)(entityName))] = (0, _deleteForm.generateDeleteForm)(generatorData);
  fs["src/routes/".concat((0, _generateFileNames.getScreenFolderName)(screenName), "/").concat((0, _generateFileNames.getTableFileName)(screenName))] = (0, _table.generateTable)(generatorData, _objectSpread({}, config, {
    includeColumns: tableFields
  }));
  fs["src/routes/".concat((0, _generateFileNames.getScreenFolderName)(screenName), "/index.js")] = (0, _generateIndex.generateIndex)(generatorData);

  if (typeof rootFile === 'string') {
    fs['src/Root.js'] = (0, _formatCode.formatCode)(rootFile.replace("".concat(PAGE_CONSTANTS.APP_PAGES_IMPORTS), "".concat(PAGE_CONSTANTS.APP_PAGES_IMPORTS).concat(os.EOL).concat(_chunks.chunks.routeImport({
      screenName: screenName,
      changeCase: changeCase
    }))).replace("".concat(PAGE_CONSTANTS.APP_ROUTE_LINKS), "".concat(PAGE_CONSTANTS.APP_ROUTE_LINKS).concat(os.EOL).concat(_chunks.chunks.routeLink({
      screenName: screenName,
      routeUrl: routeUrl,
      changeCase: changeCase
    }))).replace("".concat(PAGE_CONSTANTS.APP_ROUTES), "".concat(PAGE_CONSTANTS.APP_ROUTES).concat(os.EOL).concat(_chunks.chunks.routeComponent({
      screenName: screenName,
      routeUrl: routeUrl,
      changeCase: changeCase
    }))));
  }

  return fs;
};

exports.generateScreen = generateScreen;
