// @8base/generators/dist/generateFs/replaceInitialApp.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.replaceInitialApp = void 0;
var FS_NAMES = {
  APPLICATION_JS: 'src/Application.js',
  APPOLO_CONFIG_JS: 'apollo.config.js',
  EIGHT_BASE_YML: '.8base.yml',
  PACKAGE_JSON: 'package.json',
  PACKAGE_LOCK_JSON: 'package-lock.json'
};

var removePieceOfCode = function removePieceOfCode(code, pieceName) {
  var pieceStart = "\\/\\*\\*\\s*__".concat(pieceName, "_START__\\s*\\*\\/");
  var pieceEnd = "\\/\\*\\*\\s*__".concat(pieceName, "_END__\\s*\\*\\/");
  var regexp = new RegExp("".concat(pieceStart, "[\\s\\S]*").concat(pieceEnd), 'm');
  return code.replace(regexp, '');
};

var replaceInitialApp = function replaceInitialApp(fsObject, constants) {
  var config = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {
    authMode: 'web'
  };
  var fsObjectReplaced = {};
  var isApiTokenMode = config.authMode === 'api-token';
  Object.keys(fsObject).forEach(function (filePath) {
    switch (filePath) {
      case FS_NAMES.APPLICATION_JS:
        {
          fsObjectReplaced[FS_NAMES.APPLICATION_JS] = isApiTokenMode ? fsObject[FS_NAMES.APPLICATION_JS] && removePieceOfCode(fsObject[FS_NAMES.APPLICATION_JS], 'AUTH_WEB').replace('__APP_API_TOKEN__', constants.apiToken || '').replace('__APP_API_ENDPOINT__', constants.endpoint || '') : fsObject[FS_NAMES.APPLICATION_JS] && removePieceOfCode(fsObject[FS_NAMES.APPLICATION_JS], 'AUTH_API_TOKEN').replace('__APP_API_ENDPOINT__', constants.endpoint || '').replace('__APP_AUTH_CLIENT_ID__', constants.authClientId || '').replace('__APP_AUTH_DOMAIN__', constants.authDomain || '');
          break;
        }

      case FS_NAMES.PACKAGE_JSON:
        {
          fsObjectReplaced[FS_NAMES.PACKAGE_JSON] = fsObject[FS_NAMES.PACKAGE_JSON] && fsObject[FS_NAMES.PACKAGE_JSON].replace(/"name": ".*"/, "\"name\": \"".concat(constants.appName, "\"")).replace(/"version": ".*"/, '"version": "1.0.0"');
          break;
        }

      case FS_NAMES.PACKAGE_LOCK_JSON:
        {
          fsObjectReplaced[FS_NAMES.PACKAGE_LOCK_JSON] = fsObject[FS_NAMES.PACKAGE_LOCK_JSON] && fsObject[FS_NAMES.PACKAGE_LOCK_JSON].replace(/"name": ".*"/, "\"name\": \"".concat(constants.appName, "\"")).replace(/"version": ".*"/, '"version": "1.0.0"');
          break;
        }

      default:
        {
          fsObjectReplaced[filePath] = fsObject[filePath] && fsObject[filePath].replace('__APP_API_ENDPOINT__', constants.endpoint || '').replace('__APP_AUTH_CLIENT_ID__', constants.authClientId || '').replace('__APP_AUTH_DOMAIN__', constants.authDomain || '').replace('__APP_NAME__', constants.appName || '');
          break;
        }
    }
  });
  fsObjectReplaced[FS_NAMES.EIGHT_BASE_YML] = "appName: ".concat(constants.appName);
  return fsObjectReplaced;
};

exports.replaceInitialApp = replaceInitialApp;
