// @8base/generators/dist/generateFiles/generateCreateForm/createForm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateCreateForm = void 0;

var _schemaNameGenerator = require("@8base/schema-name-generator");

var _utils = require("@8base/utils");

var changeCase = _interopRequireWildcard(require("change-case"));

var ejs = _interopRequireWildcard(require("ejs"));

var pluralize = _interopRequireWildcard(require("pluralize"));

var _formatCode = require("../../formatCode");

var _utils2 = require("../../utils");

var _chunks = require("../chunks");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

/* babel-plugin-inline-import './createForm.js.ejs' */
// @ts-ignore
var createForm = "import React from 'react';\nimport gql from 'graphql-tag';\nimport { Query, graphql } from 'react-apollo';\nimport { Form as FormLogic, Field, FieldArray } from '@8base/forms';\nimport {\n  Dialog,\n  Grid,\n  Button,\n  Form,\n  Row,\n  Column,\n  Icon,\n  Text,\n  SelectField,\n  InputField,\n  CheckboxField,\n  DateInputField,\n  Label,\n  ModalContext,\n} from '@8base/boost';\nimport { AddressInputField, PhoneInputField, ListFields, FileInputField } from '../../shared/components'\n\nconst <%= changeCase.upper(entityName)%>_CREATE_MUTATION = gql`\n  mutation <%= changeCase.pascal(entityName)%>Create($data: <%= SchemaNameGenerator.getCreateInputName(tableName)%>!) {\n    <%= SchemaNameGenerator.getCreateItemFieldName(tableName)%>(data: $data) {\n      id\n    }\n  }\n`;\n\n<%- chunks.createEditComponents({\n    fields: fields,\n    SchemaNameGenerator: SchemaNameGenerator,\n    changeCase: changeCase,\n    chunks: chunks,\n    tableSelectors: tableSelectors,\n    table: table,\n}); %>\nconst enhancer = graphql(<%= changeCase.upper(entityName)%>_CREATE_MUTATION, {\n  name: '<%= SchemaNameGenerator.getCreateItemFieldName(tableName)%>',\n  options: {\n    refetchQueries: ['<%=changeCase.pascal(tableName)%>TableContent', '<%= changeCase.pascal(SchemaNameGenerator.getTableListFieldName(tableName))%>'],\n    context: {\n      TOAST_SUCCESS_MESSAGE: '<%= changeCase.pascal(entityName)%> successfully created',\n    },\n  },\n});\n\nconst <%= changeCase.pascal(entityName)%>CreateDialog = enhancer(\n  class <%= changeCase.pascal(entityName)%>CreateDialog extends React.PureComponent {\n    static contextType = ModalContext;\n\n    onSubmit = async data => {\n      await this.props.<%= SchemaNameGenerator.getCreateItemFieldName(tableName)%>({ variables: { data } });\n\n      this.context.closeModal('<%= changeCase.upper(entityName)%>_CREATE_DIALOG_ID');\n    };\n\n    onClose = () => {\n      this.context.closeModal('<%= changeCase.upper(entityName)%>_CREATE_DIALOG_ID');\n    };\n\n    renderFormContent = ({ handleSubmit, invalid, submitting, pristine }) => (\n      <form onSubmit={handleSubmit}>\n        <Dialog.Header title=\"New <%= changeCase.pascal(entityName)%>\" onClose={this.onClose} />\n        <Dialog.Body scrollable>\n          <Grid.Layout gap=\"md\" stretch><%-\n            chunks.fieldsInputsList({ \n              fields: fields, \n              chunks: chunks,\n              changeCase: changeCase,\n              pluralize: pluralize,\n              SchemaNameGenerator: SchemaNameGenerator,\n            });%>\n          </Grid.Layout>\n        </Dialog.Body>\n        <Dialog.Footer>\n          <Button color=\"neutral\" type=\"button\" variant=\"outlined\" disabled={submitting} onClick={this.onClose}>\n            Cancel\n          </Button>\n          <Button color=\"primary\" type=\"submit\" loading={submitting}>\n            Create <%= changeCase.pascal(entityName)%>\n          </Button>\n        </Dialog.Footer>\n      </form>\n    );\n\n    render() {\n      return (\n        <Dialog id={'<%= changeCase.upper(entityName)%>_CREATE_DIALOG_ID'} size=\"sm\">\n          <FormLogic type=\"CREATE\" tableSchemaName=\"<%= tableName%>\" onSubmit={this.onSubmit} formatRelationToIds>\n            {this.renderFormContent}\n          </FormLogic>\n        </Dialog>\n      );\n    }\n  }\n);\n\nexport { <%= changeCase.pascal(entityName)%>CreateDialog };\n";

var generateCreateForm = function generateCreateForm(_ref) {
  var tablesList = _ref.tablesList,
      tableId = _ref.tableId,
      screenName = _ref.screenName;

  var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      includeColumns = _ref2.includeColumns;

  var table = tablesList.find(function (_ref3) {
    var id = _ref3.id;
    return tableId === id;
  });

  if (!table) {
    throw new Error("Can't find a table with ".concat(tableId, " id"));
  }

  var tableName = table.displayName || table.name;
  var entityName = pluralize.singular(tableName);
  var fields = table.fields.filter(function (_ref4) {
    var isMeta = _ref4.isMeta,
        name = _ref4.name;
    return !isMeta && (0, _utils2.isFieldNeedsToInclude)(name, includeColumns);
  });
  var tableGenerated = ejs.render(createForm, {
    SchemaNameGenerator: _schemaNameGenerator.SchemaNameGenerator,
    changeCase: changeCase,
    chunks: _chunks.chunks,
    entityName: entityName,
    fields: fields,
    pluralize: pluralize,
    screenName: screenName || entityName,
    table: table,
    tableName: tableName,
    tableSelectors: _utils.tableSelectors
  });
  return (0, _formatCode.formatCode)(tableGenerated);
};

exports.generateCreateForm = generateCreateForm;
