// @8base/utils/dist/selectors/tableFieldSelectors.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const R = __importStar(require("ramda"));
const reselect_1 = require("reselect");
const constants_1 = require("../constants");
exports.getTableField = (tableField) => tableField;
exports.getFieldType = reselect_1.createSelector(exports.getTableField, R.prop('fieldType'));
exports.getFieldTypesAttributes = reselect_1.createSelector(exports.getTableField, R.prop('fieldTypeAttributes'));
exports.isRelationField = reselect_1.createSelector(exports.getFieldType, R.equals(constants_1.FIELD_TYPE.RELATION));
exports.isFileField = reselect_1.createSelector(exports.getFieldType, R.equals(constants_1.FIELD_TYPE.FILE));
exports.isSmartField = reselect_1.createSelector(exports.getFieldType, R.equals(constants_1.FIELD_TYPE.SMART));
exports.isIdField = reselect_1.createSelector(exports.getFieldType, R.equals(constants_1.FIELD_TYPE.ID));
exports.isMetaField = reselect_1.createSelector(exports.getTableField, R.propEq('isMeta', true));
exports.isListField = reselect_1.createSelector(exports.getTableField, R.propEq('isList', true));
exports.isSystemField = reselect_1.createSelector(exports.getTableField, R.propEq('isSystem', true));
exports.getFieldName = reselect_1.createSelector(exports.getTableField, R.prop('name'));
exports.getRelationTableId = reselect_1.createSelector(exports.getTableField, R.path(['relation', 'refTable', 'id']));
exports.getRelationTableName = reselect_1.createSelector(exports.getTableField, R.path(['relation', 'refTable', 'name']));
