// @8base/generators/dist/generateFiles/generateEditForm/editForm.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.generateEditForm = void 0;

var _schemaNameGenerator = require("@8base/schema-name-generator");

var _utils = require("@8base/utils");

var changeCase = _interopRequireWildcard(require("change-case"));

var ejs = _interopRequireWildcard(require("ejs"));

var pluralize = _interopRequireWildcard(require("pluralize"));

var _formatCode = require("../../formatCode");

var _utils2 = require("../../utils");

var _chunks = require("../chunks");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

/* babel-plugin-inline-import './editForm.js.ejs' */
// @ts-ignore
var editForm = "import React from 'react';\nimport gql from 'graphql-tag';\nimport { Query, graphql } from 'react-apollo';\nimport { Form as FormLogic, Field, FieldArray } from '@8base/forms';\nimport {\n  AsyncContent,\n  Dialog,\n  Grid,\n  Button,\n  Form,\n  Row,\n  Column,\n  Icon,\n  Text,\n  SelectField,\n  InputField,\n  CheckboxField,\n  DateInputField,\n  Label,\n  ModalContext,\n} from '@8base/boost';\nimport { AddressInputField, PhoneInputField, ListFields, FileInputField } from '../../shared/components'\n\nconst <%= changeCase.upper(entityName)%>_QUERY = gql`\n  <%= queryEntityText%> \n`;\n\nconst <%= changeCase.upper(entityName)%>_UPDATE_MUTATION = gql`\n  mutation <%= changeCase.pascal(entityName)%>Update($data: <%= SchemaNameGenerator.getUpdateInputName(tableName)%>!) {\n    <%= SchemaNameGenerator.getUpdateItemFieldName(tableName)%>(data: $data) {\n      id\n    }\n  }\n`;\n\n<%- chunks.createEditComponents({\n    fields: fields,\n    SchemaNameGenerator: SchemaNameGenerator,\n    changeCase: changeCase,\n    chunks: chunks,\n    tableSelectors: tableSelectors,\n    table: table,\n}); %>\n\nconst ehnhancer = graphql(<%= changeCase.upper(entityName)%>_UPDATE_MUTATION, {\n  name: '<%= SchemaNameGenerator.getUpdateItemFieldName(tableName)%>',\n  options: {\n    refetchQueries: ['<%=changeCase.pascal(tableName)%>TableContent', '<%= changeCase.pascal(SchemaNameGenerator.getTableListFieldName(tableName))%>'],\n    context: {\n      TOAST_SUCCESS_MESSAGE: '<%= changeCase.pascal(entityName)%> successfully updated',\n    },\n  },\n});\n\nconst <%= changeCase.pascal(entityName)%>EditDialog = ehnhancer(\n  class <%= changeCase.pascal(entityName)%>EditDialog extends React.PureComponent {\n    static contextType = ModalContext;\n\n    updateOnSubmit = id => async data => {\n      await this.props.<%= SchemaNameGenerator.getUpdateItemFieldName(tableName)%>({ variables: { data: { ...data, id } } });\n\n      this.context.closeModal('<%= changeCase.upper(entityName)%>_EDIT_DIALOG_ID');\n    };\n\n    onClose = () => {\n      this.context.closeModal('<%= changeCase.upper(entityName)%>_EDIT_DIALOG_ID');\n    };\n\n    renderForm = ({ args }) => {\n      return (\n        <Query query={<%= changeCase.upper(entityName)%>_QUERY} variables={{ id: args.id }}>\n          {({ data, loading }) => (\n          <FormLogic\n            type=\"UPDATE\"\n            tableSchemaName=\"<%= tableName%>\"\n            onSubmit={this.updateOnSubmit(args.id)}\n            initialValues={data.<%= changeCase.lower(entityName) %>}\n            formatRelationToIds\n          >\n            {({ handleSubmit, invalid, submitting, pristine }) => (\n              <form onSubmit={handleSubmit}>\n                <Dialog.Header title=\"Edit <%= changeCase.pascal(entityName)%>\" onClose={this.onClose} />\n                <Dialog.Body scrollable>\n                  <AsyncContent loading={ loading } stretch>\n                    <Grid.Layout gap=\"md\" stretch>\n                      <%-\n                            chunks.fieldsInputsList({ \n                              fields: fields,\n                              chunks: chunks,\n                              changeCase: changeCase,\n                              pluralize: pluralize,\n                              SchemaNameGenerator: SchemaNameGenerator,\n                            }); %>\n                    </Grid.Layout>\n                  </AsyncContent>\n                </Dialog.Body>\n                <Dialog.Footer>\n                  <Button color=\"neutral\" type=\"button\" variant=\"outlined\" disabled={submitting} onClick={this.onClose}>\n                    Cancel\n                  </Button>\n                  <Button color=\"primary\" type=\"submit\" disabled={pristine || invalid} loading={submitting}>\n                    Update\n                    <%= changeCase.pascal(entityName)%>\n                  </Button>\n                </Dialog.Footer>\n              </form>\n              )}\n          </FormLogic>\n        )}\n        </Query>\n      );\n    };\n\n    render() {\n      return (\n        <Dialog id={'<%= changeCase.upper(entityName)%>_EDIT_DIALOG_ID'} size=\"sm\">\n          {this.renderForm}\n        </Dialog>\n      );\n    }\n  }\n);\n\nexport { <%= changeCase.pascal(entityName)%>EditDialog };\n";

var generateEditForm = function generateEditForm(_ref) {
  var tablesList = _ref.tablesList,
      tableId = _ref.tableId,
      screenName = _ref.screenName;

  var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      includeColumns = _ref2.includeColumns;

  var table = tablesList.find(function (_ref3) {
    var id = _ref3.id;
    return tableId === id;
  });

  if (!table) {
    throw new Error("Can't find a table with ".concat(tableId, " id"));
  }

  var tableName = table.displayName || table.name;
  var entityName = pluralize.singular(screenName || tableName);
  var mutationText = (0, _utils.createTableRowUpdateTag)(tablesList, tableId);
  var queryEntityText = (0, _utils.createTableRowQueryTag)(tablesList, tableId, {
    withMeta: false
  });
  var fields = table.fields.filter(function (_ref4) {
    var isMeta = _ref4.isMeta,
        name = _ref4.name;
    return !isMeta && (0, _utils2.isFieldNeedsToInclude)(name, includeColumns);
  });
  var tableGenerated = ejs.render(editForm, {
    SchemaNameGenerator: _schemaNameGenerator.SchemaNameGenerator,
    changeCase: changeCase,
    chunks: _chunks.chunks,
    entityName: entityName,
    fields: fields,
    mutationText: mutationText,
    pluralize: pluralize,
    queryEntityText: queryEntityText,
    screenName: screenName || tableName,
    table: table,
    tableName: tableName,
    tableSelectors: _utils.tableSelectors
  });
  return (0, _formatCode.formatCode)(tableGenerated);
};

exports.generateEditForm = generateEditForm;
